pkg/lib/synthesize/generate.go

```diff
@@ -0,0 +1,174 @@
+package synthesize
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/yaml"
+
+	"github.com/operator-framework/operator-registry/pkg/lib/bundle"
+)
+
+var CsvInferenceFileNotFoundError = errors.New("a file that contains non-inferable metadata could not be found")
+
+const (
+	SynthesizedCSVFileName string = "olm.clusterserviceversion.yaml"
+	DefaultPermission             = 0644
+)
+
+// GenerateCSV generates a CSV for the bundle based on the resource within the bundle directory.
+func GenerateCSV(bundleDir string) error {
+	bundleDir, err := filepath.Abs(bundleDir)
+	if err != nil {
+		return err
+	}
+
+	// find annotations.yaml file in the bundle dir.
+	var annotations bundle.AnnotationMetadata
+	var annotationsFilePath string
+	if err := filepath.Walk(bundleDir, func(path string, info os.FileInfo, err error) error {
+		if annotations.IsComplete() {
+			return nil
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		annotationsFilePath = path
+		content, err := ioutil.ReadFile(annotationsFilePath)
```

Comment:
Will this blow up if a very large file is encountered while walking the tree? Might save a headache later to use a https://pkg.go.dev/io?tab=doc#LimitReader. Accepting an io.Reader instead of []byte in GetAnnotations would also let reading bail out early (as soon as the decoder recognizes that it's not YAML, for example) instead of doing a bunch of unnecessary I/O.